### DISCUSSION TITLE
Taking a lot of time to load from Cog Titiler due to huge number of bands

### DISCUSSION THREAD

**Discussion by sairahul1526** [developmentseed/titiler#958] [Q&A] [answered]
Download of cog file is faster but taking a lot of time to visualise if there are huge number of bands in it.

Is there any way to speed it up or use multi processing for this?

Below are the details of the COG

`Size is 677, 462
Coordinate System is:
GEOGCRS["WGS 84",
  ENSEMBLE["World Geodetic System 1984 ensemble",
    MEMBER["World Geodetic System 1984 (Transit)"],
    MEMBER["World Geodetic System 1984 (G730)"],
    MEMBER["World Geodetic System 1984 (G873)"],
    MEMBER["World Geodetic System 1984 (G1150)"],
    MEMBER["World Geodetic System 1984 (G1674)"],
    MEMBER["World Geodetic System 1984 (G1762)"],
    MEMBER["World Geodetic System 1984 (G2139)"],
    MEMBER["World Geodetic System 1984 (G2296)"],
    ELLIPSOID["WGS 84",6378137,298.257223563,
      LENGTHUNIT["metre",1]],
    ENSEMBLEACCURACY[2.0]],
  PRIMEM["Greenwich",0,
    ANGLEUNIT["degree",0.0174532925199433]],
  CS[ellipsoidal,2],
    AXIS["geodetic latitude (Lat)",north,
      ORDER[1],
      ANGLEUNIT["degree",0.0174532925199433]],
    AXIS["geodetic longitude (Lon)",east,
      ORDER[2],
      ANGLEUNIT["degree",0.0174532925199433]],
  USAGE[
    SCOPE["Horizontal component of 3D system."],
    AREA["World."],
    BBOX[-90,-180,90,180]],
  ID["EPSG",4326]]
Data axis to CRS axis mapping: 2,1
Origin = (38.280810923857416,30.063159212119068)
Pixel Size = (0.000270293649014,-0.000270179236462)
Metadata:
 AREA_OR_POINT=Area
Image Structure Metadata:
 LAYOUT=COG
 COMPRESSION=LZW
 INTERLEAVE=PIXEL
Corner Coordinates:
Upper Left ( 38.2808109, 30.0631592) ( 38d16'50.92"E, 30d 3'47.37"N)
Lower Left ( 38.2808109, 29.9383364) ( 38d16'50.92"E, 29d56'18.01"N)
Upper Right ( 38.4637997, 30.0631592) ( 38d27'49.68"E, 30d 3'47.37"N)
Lower Right ( 38.4637997, 29.9383364) ( 38d27'49.68"E, 29d56'18.01"N)
Center   ( 38.3723053, 30.0007478) ( 38d22'20.30"E, 30d 0' 2.69"N)
Band 1 Block=512x512 Type=UInt16, ColorInterp=Gray
 NoData Value=0
 Overviews: 338x231
Band 2 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 3 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 4 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 5 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 6 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 7 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 8 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 9 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 10 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 11 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 12 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 13 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 14 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 15 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 16 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 17 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 18 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 19 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 20 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 21 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 22 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 23 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 24 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 25 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 26 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 27 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 28 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 29 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 30 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 31 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 32 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 33 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 34 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 35 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 36 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 37 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 38 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 39 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 40 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 41 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 42 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 43 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 44 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 45 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 46 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 47 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 48 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 49 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 50 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 51 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 52 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 53 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 54 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 55 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 56 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 57 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 58 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 59 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 60 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 61 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 62 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 63 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 64 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231
Band 65 Block=512x512 Type=UInt16, ColorInterp=Undefined
 NoData Value=0
 Overviews: 338x231`

**Comment by vincentsarago** [comment]
> Download of cog file is faster but taking a lot of time to visualise if there are huge number of bands in it.

Pretty hard to tell without knowing how you're trying to render your file 🤷‍♂️

**Reply by sairahul1526** [reply]
we are using GET cog tiles endpoint to render and visualise on client - https://titiler.xyz/api.html#/Cloud%20Optimized%20GeoTIFF/tile_cog_tiles__tileMatrixSetId___z___x___y__get

**Reply by sairahul1526** [reply]
any idea @vincentsarago

**Reply by vincentsarago** [reply]
not without the actual file nor tile request parameters 🤷

**Reply by dimplejaiin** [reply]
@vincentsarago, these are the tile request parameters we are using  - 
`https://stacapi/cog/cog/preview.jpg?url=<s3tifurl>&resampling_method=nearest&bidx=37&bidx=14&bidx=1&rescale=0%2C50000&max_size=200`

**Reply by dimplejaiin** [reply]
@vincentsarago Right now, I'm unable to share the file here, but please let us know what specific information or details you need from the raster.

**Reply by vincentsarago** [reply]
you can use https://github.com/cogeotiff/rio-faux to create a fake version of your file

**Reply by dimplejaiin** [reply]
Thank you, @vincentsarago, for your quick response. I have added two copies of the [files](https://drive.google.com/drive/folders/1ZRN4vMeJ8KPjKaRatWn9UMm1nNkn66u2?usp=sharing): one with a fake version and the other with noise. Please have a look at the files and let us know if you need anything else.

About the endpoint and request parameters we are using:
- `https://stacapi/cog/cog/preview.jpg?url=<s3_tif>&resampling_method=nearest&bidx=37&bidx=14&bidx=1&rescale=0%2C50000&max_size=200`
- `https://api.pixxel.space/cog/cog/tiles/WebMercatorQuad/z/x/y.webp?url=<s3tif>&resampling_method=nearest&bidx=4&bidx=3&bidx=2&rescale=0%2C5000`

**Reply by vincentsarago** [reply]
@dimplejaiin thanks for providing the files.

TBH I don't see anything `slow`, could you explain what you meant by `slow`? Do you have specific timings? 

you can also use https://github.com/developmentseed/tilebench?tab=readme-ov-file#api to get more `timings` info

**Reply by dimplejaiin** [reply]
@vincentsarago here are the specific metrics:
- Current speed for loading shared tile: 43.66371202468872 seconds.
- Loading COG tiles from Sentinel-2 takes less than a second.
- The pod memory usage details:
![Screenshot 2024-08-01 at 4 42 41 PM](https://github.com/user-attachments/assets/91deb15d-12e0-4d59-a21b-7aff3cc672fd)

**Reply by dimplejaiin** [reply]
Hi @vincentsarago ,
I have conducted performance profiling using `tilebench` for Sentinel-2 and dataset where we are experiencing slow rendering [file](https://drive.google.com/drive/folders/1ZRN4vMeJ8KPjKaRatWn9UMm1nNkn66u2?usp=sharing). Here are the detailed timings:

#### Timings Information for dataset where we are experiencing slow rendering:
```bash
###Get Zoom Levels:
$ tilebench get-zooms <s3url>
{"minzoom": 11, "maxzoom": 12}

###Tile profile:
$ tilebench profile <s3url>
2024-08-06T13:42:24.150651+0530 | TILEBENCH | reading tile: 11-1480-928
{"HEAD": {"count": 0}, "GET": {"count": 6, "bytes": 35646127, "ranges": ["0-16383", "16384-1064959", "1064960-11550719", "11550720-27934719", "27934720-35646126"]}, "Timing": 6.003262042999268}

###Random Tile at Zoom 15:
$ tilebench random <s3url> --zoom 15
15-23679-14861

###Profile Specific Tile at Zoom 15 with Configurations:
$ tilebench profile <s3url> --tile 15-7055-13040 --config GDAL_DISABLE_READDIR_ON_OPEN=EMPTY_DIR    
{"HEAD": {"count": 0}, "GET": {"count": 5, "bytes": 35629743, "ranges": ["16384-1064959", "1064960-11550719", "11550720-27934719", "27934720-35646126"]}, "Timing": 6.925491809844971}

$ tilebench profile <s3url> --tile 15-7055-13040 --config GDAL_DISABLE_READDIR_ON_OPEN=FALSE   
{"HEAD": {"count": 0}, "GET": {"count": 6, "bytes": 35646127, "ranges": ["0-16383", "16384-1064959", "1064960-11550719", "11550720-27934719", "27934720-35646126"]}, "Timing": 6.576103925704956}

```

#### Sentinel-2:
```bash
###Get Zoom Levels:
$ tilebench get-zooms <s3url>
{"minzoom": 9, "maxzoom": 14}

###Tile profile:
$ tilebench profile <s3url>
2024-08-06T12:57:25.225692+0530 | TILEBENCH | reading tile: 10-219-406
{"HEAD": {"count": 0}, "GET": {"count": 4, "bytes": 376832, "ranges": ["0-16383", "180224-540671"]}, "Timing": 1.7954511642456055}

###Random Tile at Zoom 15:
$ tilebench random <s3url> --zoom 15
15-7055-13040

###Profile Specific Tile at Zoom 15 with Configurations:
$ tilebench profile <s3url> --tile 15-7055-13040 --config GDAL_DISABLE_READDIR_ON_OPEN=EMPTY_DIR    
{"HEAD": {"count": 0}, "GET": {"count": 3, "bytes": 49152, "ranges": ["13975552-13991935", "14024704-14057471"]}, "Timing": 1.4548020362854004}

$ tilebench profile <s3url> --tile 15-7055-13040 --config GDAL_DISABLE_READDIR_ON_OPEN=FALSE   
{"HEAD": {"count": 0}, "GET": {"count": 4, "bytes": 65536, "ranges": ["0-16383", "13975552-13991935", "14024704-14057471"]}, "Timing": 1.756836175918579}
```

Could you review this and provide any insights or recommendations for optimization? Please let me know if anything else is required from my end.

**Reply by vincentsarago** [reply]
well to me it seems that the connection to S3 is super slow

6 sec for 6 GET seems a bit weird to me

**Reply by punit-kulal** [reply, marked as answer]
The COG created by GDAL is INTERLEAVED=PIXEL, hence fetching a subset of band indexes isn't supported. Hence, even if we query it, with  a subset of the bands at high level and pass it to rasterio. (Rasterio needs to explicitly download an entire overview containing all bands, which covers the tile) and then extract the bands from it. 

Ref:
- https://gdal.org/drivers/raster/cog.html#high-level
- https://github.com/OSGeo/gdal/issues/4931

Note: The COG specification itself is unopinionated about it as mentioned [here](https://github.com/opengeospatial/CloudOptimizedGeoTIFF/issues/3).

**Reply by vincentsarago** [reply]
This makes senses 🙏 

When dealing with a high number of bands, having `INTERLEAVED=PIXEL` does mean that GDAL will fetch the whole set of bands before to be able to extract individual band value 

> If you are including all band on a Landsat in a single COG file, the common thing to do is extract and show a few band but rarely all of them so PlanarConfiguration=Separate seems more appropriate.

from https://github.com/opengeospatial/CloudOptimizedGeoTIFF/issues/3#issuecomment-987133206
